drivers: pwm: nrf_sw: support generating PWM based on RTC

mniestroj opened the pull request:
So far nRF's TIMER was used for generating PWM signal. Add support for
generating PWM based on RTC, which is sourced by 32KHz low frequency
crystal. This allows to use low frequency PWM with much lower power
consumption, because high frequency clock path can be disabled.

Tested with RTC0 and RTC2, with `channel-count` equal 1 or 2.

mniestroj committed:
drivers: pwm: nrf_sw: support generating PWM based on RTC

So far nRF's TIMER was used for generating PWM signal. Add support for
generating PWM based on RTC, which is sourced by 32KHz low frequency
crystal. This allows to use low frequency PWM with much lower power
consumption, because high frequency clock path can be disabled.

Don't support RTC clock prescaler, because maximum 512s period covers
most use cases. This allows to adjust pulse and period cycles to the
fact that CLEAR task event is generated always one LFCLK cycle after
period COMPARE value is reached.

Also update hal_nordic revision, as it contains updated check for PPI
channels conflict when RTC is used to generate PWM.

Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>

anangl commented on dts/bindings/pwm/nordic,nrf-sw-pwm.yaml:
```diff
@@ -5,10 +5,15 @@ compatible: "nordic,nrf-sw-pwm"
 properties:
+    rtc-instance:
+      type: int
+      description: RTC instance to use for generating the PWM output signals
```
Descriptions of `channel-count`: https://github.com/zephyrproject-rtos/zephyr/blob/99f5fc4d21572ff3ff177a36683fb339364ec378/dts/bindings/pwm/nordic%2Cnrf-sw-pwm.yaml#L20
and `clock-prescaler`: https://github.com/zephyrproject-rtos/zephyr/blob/99f5fc4d21572ff3ff177a36683fb339364ec378/dts/bindings/pwm/nordic%2Cnrf-sw-pwm.yaml#L25
should be also updated.

anangl reviewed (changes requested):
The following description needs to be also updated: https://github.com/zephyrproject-rtos/zephyr/blob/99f5fc4d21572ff3ff177a36683fb339364ec378/drivers/pwm/Kconfig.nrf5_sw#L12-L13
And [this definition](https://github.com/zephyrproject-rtos/hal_nordic/blob/fc12f5e2b47cf143254234c05f9ecbdb603a7101/nrfx_glue.h#L260-L262) as well:
```
#define NRFX_PPI_CHANNELS_USED_BY_PWM_SW \
    (BIT_MASK(DT_PROP(PWM_NRF5_SW_NODE, channel_count) * 2) \
         << DT_PROP(PWM_NRF5_SW_NODE, ppi_base))
```

mniestroj commented:
```
-:51: ERROR:SPACING: need consistent spacing around '*' (ctx:WxV)
#51: FILE: drivers/pwm/pwm_nrf5_sw.c:41:
+	NRF_RTC_Type *rtc;
```
seems like a false positive. Any idea why checkpatch is complaining? After changing to `NRF_RTC_Type * rtc;` it stops warning us, but that is not the style used in Zephyr.

anangl commented on drivers/pwm/pwm_nrf5_sw.c:
```diff
@@ -160,31 +199,58 @@ static int pwm_nrf5_sw_pin_set(const struct device *dev, uint32_t pwm,
 	/* calc div, to scale down to fit in 16 bits */
 	div = period_cycles >> 16;
```
RTC has 24-bit compare registers.
But this "scaling" is anyway incorrect, even for the 16-bit TIMER case. Instead, something like this should be done above (for RTC, the TIMER case can be corrected in a separate PR):
```
if (USE_RTC && period_cycles > BITMASK(24)) {
	LOG_ERR("Too long period (%u)", period_cycles);
	return -EINVAL;
}
```

mniestroj replied:
Am I right then, that this scaling (`div = period_cycles >> 16;`) should be dropped entirely for both RTC and TIMER?

anangl replied:
I think so. Currently it just uses incorrect settings, and thus gives incorrect PWM output. It will be better to return an error with some hint to the user, like in the mcux driver: https://github.com/zephyrproject-rtos/zephyr/blob/3c79b565fe7e6afed3687d95dd3e70bb93699a88/drivers/pwm/pwm_mcux.c#L58-L64

mniestroj replied:
Added similar check for both TIMER and RTC.

anangl commented:
> ```
> -:51: ERROR:SPACING: need consistent spacing around '*' (ctx:WxV)
> #51: FILE: drivers/pwm/pwm_nrf5_sw.c:41:
> +	NRF_RTC_Type *rtc;
> ```
> 
> seems like a false positive. Any idea why checkpatch is complaining? After changing to `NRF_RTC_Type * rtc;` it stops warning us, but that is not the style used in Zephyr.

Looks like a false positive, indeed.

mniestroj commented on drivers/pwm/Kconfig.nrf5_sw:
```diff
@@ -9,5 +9,9 @@ config PWM_NRF5_SW
+	  When used with HF timer it provides up to 3 pins, utilizing two PPI
+	  channels per pin and one GPIOTE config per pin.
+
+	  When used with RTC it provides up to 2 pins, utilizing three PPI
```
My mistake, I have confused PPI with RTC's CC. It should be 3 of course.

mniestroj commented on drivers/pwm/pwm_nrf5_sw.c:
```diff
@@ -160,31 +199,58 @@ static int pwm_nrf5_sw_pin_set(const struct device *dev, uint32_t pwm,
+			(uint32_t) &(rtc->EVENTS_COMPARE[config->map_size]);
+		NRF_PPI->CH[ppi_index + 1].TEP =
+			(uint32_t) &(NRF_GPIOTE->TASKS_OUT[channel]);
+		NRF_PPI->CH[ppi_index + 2].EEP =
+			(uint32_t) &(rtc->EVENTS_COMPARE[config->map_size]);
+		NRF_PPI->CH[ppi_index + 2].TEP =
+			(uint32_t) &(rtc->TASKS_CLEAR);
```
Seems reasonable to drop prescaling support, I cannot imagine use case for >512s PWM period.

So I guess we should two things:
1) force `clock-prescaler = <0>;` with `BUILD_ASSERT()`,
2) drop `div = period_cycles >> 16;` calculation entirely for RTC (the same can be done for TIMER) and just check that cycles value fit in 24 bits.

mniestroj replied:
I've added commit which drops `div = period_cycles >> 16;` for TIMER. It just seemed more logical to drop broken mechanism, even in this PR, before introducing new features.

Additionally I've adjusted period_cycles and pulse_cycles to take one LFCLK cycle delay:
```
	rtc->CC[channel] = pulse_cycles - 1;
	rtc->CC[config->map_size] = period_cycles - 1;
```

mniestroj committed:
drivers: pwm: nrf_sw: drop broken runtime timer clock scaling

There is some kind of runtime clock scaling implemented when period
cycles exceed 16-bit value. This is broken, because configured PWM has
far higher frequency in that case. Replace that mechanism with simply
veryfying if requested PWM period does not exceed 16-bit value.

Suggested-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>
Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>

mbolivar-nordic commented on dts/bindings/pwm/nordic,nrf-sw-pwm.yaml:
```diff
@@ -5,19 +5,32 @@ compatible: "nordic,nrf-sw-pwm"
 
 properties:
+    rtc-instance:
+      type: int
```
Why not phandle and pointing to the `nordic,nrf-rtc` directly? I think it should probably be done that way, just looking at this quickly.

mbolivar-nordic commented on dts/bindings/pwm/nordic,nrf-sw-pwm.yaml:
```diff
@@ -5,19 +5,32 @@ compatible: "nordic,nrf-sw-pwm"
+      required: false
+
     timer-instance:
       type: int
```
I see that this already uses an instance number, but I think it would be more appropriate to use phandles.

I haven't reviewed the code in detail, though. I am not putting a -1 or a +1 as a result, just offering a quick comment. I will review further when I can, just wanted to have something to turn around for the time zone difference.

mniestroj replied:
I agree, that using phandles is better. The reason why I went with `rtc-instance` is that I wanted to solve one issue at a time, which is adding RTC support to generate PWM. If such conversion to phandle is straightforward to do and you can point to some other driver which resolves RTC/TIMER instance number based on phandle, then I can update this PR. If that is not so simple, then I prefer to convert `rtc-instance` (`timer-instance`) integer to `rtc` (`timer`) phandle in another PR.

mbolivar-nordic commented:
Marcin Niestroj <notifications@github.com> writes:

> @mniestroj commented on this pull request.
>
>
>
>> @@ -5,19 +5,32 @@ compatible: "nordic,nrf-sw-pwm"
>  include: [pwm-controller.yaml, base.yaml]
>  
>  properties:
> +    rtc-instance:
> +      type: int
> +      description: RTC instance to use for generating the PWM output signals
> +      required: false
> +
>      timer-instance:
>        type: int
>
> I agree, that using phandles is better. The reason why I went with
> `rtc-instance` is that I wanted to solve one issue at a time, which is
> adding RTC support to generate PWM. If such conversion to phandle is
> straightforward to do and you can point to some other driver which
> resolves RTC/TIMER instance number based on phandle, then I can update
> this PR. If that is not so simple, then I prefer to convert
> `rtc-instance` (`timer-instance`) integer to `rtc` (`timer`) phandle
> in another PR.

I wasn't able to think of anything so far, so let's just do it your way.

>
> -- 
> You are receiving this because your review was requested.
> Reply to this email directly or view it on GitHub:
> https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fzephyrproject-rtos%2Fzephyr%2Fpull%2F29094%23discussion_r513266243&amp;data=04%7C01%7Cmarti.bolivar%40nordicsemi.no%7C45ad6bbaee734784e3b808d87b1d80af%7C28e5afa2bf6f419a8cf6b31c6e9e5e8d%7C0%7C0%7C637394713925109942%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&amp;sdata=Q8JUTHlCUQCkmjNw%2B4kbZ%2B0BVSMHHSPaLbOkWoN3cNU%3D&amp;reserved=0

pabigot commented:
> I wasn't able to think of anything so far, so let's just do it your way.

I had a little time left for the month and also want to get rid of instance numbers so I gave it a shot.

A potential solution is in the top commit [here](https://github.com/pabigot/zephyr/commits/pr/29094).  It seems pretty clean, but the fact the Nordic RTC and TIMER devicetree nodes **still** don't provide the number of CC registers associated with the instance requires eliminating a compile-time check.

I've done build and value tests of these on samples/boards/bbc_microbit/sound, but I can't confirm the driver actually works because I don't have a buzzer to attach.

You're welcome to integrate this or not, as you wish.

mniestroj commented:
@pabigot I really like your proposal. That way `rtc-instance` doesn't have to be introduced at all. Let's wait for `cc-num` PR and I will update this PR then.

mniestroj committed:
drivers: pwm: nrf_sw: convert to timer phandle instead instance number

So far 'timer-instance' DT property was the way to configure TIMER which
was used for generating PWM signal. Replace that by using 'generator'
phandle, so we get prepared for supporting RTC instances as PWM signal
generator.

Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>

mniestroj commented:
There is one additional commit, that switches from timer instance number to timer phandle. Adding RTC support got simplified a little bit, so we got less preprocessor usage at the beginning of file.

anangl commented on drivers/pwm/pwm_nrf5_sw.c:
```diff
@@ -153,38 +205,59 @@ static int pwm_nrf5_sw_pin_set(const struct device *dev, uint32_t pwm,
 
-	timer->EVENTS_COMPARE[channel] = 0;
-	timer->EVENTS_COMPARE[config->map_size] = 0;
+		rtc->CC[channel] = pulse_cycles - 1;
+		rtc->CC[config->map_size] = period_cycles - 1;
```
Maybe it would be good to add a comment why these `- 1` are needed here.

mniestroj replied:
Added a comment about it.

mniestroj replied:
Added a comment about it.

anangl commented on drivers/pwm/pwm_nrf5_sw.c:
```diff
@@ -15,17 +15,30 @@
 #define PWM_0_MAP_SIZE DT_INST_PROP(0, channel_count)
 
 struct pwm_config {
+#if USE_RTC
+	NRF_RTC_Type *rtc;
+#else
 	NRF_TIMER_Type *timer;
+#endif
```
How about using a union here? Maybe this could please the compliance check?

mniestroj replied:
Indeed, using `union` helped with `checkpatch.pl`. `#ifdef` might have confused it somehow.

mbolivar-nordic commented on dts/bindings/pwm/nordic,nrf-sw-pwm.yaml:
```diff
@@ -5,9 +5,9 @@ compatible: "nordic,nrf-sw-pwm"
-      description: Timer instance to use for generating the PWM output signals
+    generator:
+      type: phandle
+      description: Reference to TIMER instance for generating PWM output signals
```
This will break compatibility with any code that's using timer-instance.

Do you think you could fall back on `timer-instance` if `generator` is not defined?

Something like this:

```c
#if DT_INST_NODE_HAS_PROP(0, timer_instance) && DT_INST_NODE_HAS_PROP(0, generator)
#error "Please define either the timer-instance or generator property, but not both"
#elif DT_INST_NODE_HAS_PROP(0, timer_instance)
#define GENERATOR_ADDR _CONCAT(NRF_TIMER, DT_INST_PROP(0, timer_instance))
#else
#define GENERATOR_ADDR	((NRF_TIMER_Type *) DT_REG_ADDR(GENERATOR_NODE))
#endif
```

If you can put in a comment saying 'timer-instance' is deprecated, that would be useful.

pabigot replied:
Didn't we just get support for deprecated properties from #29417?  We could support both for a while, right?

If `timer-instance` is removed from the binding I wouldn't expect the conditional code to work.

mbolivar-nordic replied:
Sorry if I was unclear.

I am suggesting we do **not** remove this from the binding, exactly so the conditional code will work.

mniestroj replied:
I have updated code to keep `timer-instance` as deprecated property. Thanks for all suggestions!

mbolivar-nordic reviewed (approved):
Nice work!

carlescufi commented:
@mniestroj please update to SHA

anangl commented on west.yml:
```diff
@@ -56,7 +56,7 @@ manifest:
       path: modules/hal/infineon
     - name: hal_nordic
-      revision: 7430036d4f18cdf1f49158e97e5d57aacb44fb92
+      revision: pull/61/head
```
@mniestroj Please update to:
```suggestion
      revision: 031065e8f9066b4e5fcda57f390af1cd67f35a65
```
as https://github.com/zephyrproject-rtos/hal_nordic/pull/61 got merged.
